lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -1,17 +1,1035 @@
+"""Tests for the Tool Meister module.
 """
-Tests for the Tool Meister modules.
-"""
+
+from http import HTTPStatus
 import io
 import logging
-import shutil
-import subprocess
-from http import HTTPStatus
 from pathlib import Path
-
 import pytest
 import responses
+import shutil
+import signal
+import subprocess
+from typing import Any, List, NamedTuple, Tuple
+
+from pbench.agent.tool_meister import (
+    log_raw_io_output,
+    Tool,
+    TransientTool,
+    PcpTransientTool,
+    PersistentTool,
+    DcgmTool,
+    NodeExporterTool,
+    PcpTool,
+    ToolException,
+    ToolMeister,
+    ToolMeisterError,
+)
+
+
+def test_log_raw_io_output(caplog):
+    class MockedPipe:
+        def readlines(self, hint: int = -1) -> bytes:
+            yield b"line1\n"
+            yield b"\n"
+            yield b"line3\n"
+            yield b"line4\n"
+
+    log_raw_io_output(MockedPipe(), logging.getLogger("test_log_raw_io_output"))
+
+    assert caplog.record_tuples[0][2] == "line1"
+    assert caplog.record_tuples[1][2] == "line3"
+    assert caplog.record_tuples[2][2] == "line4"
+
+
+class NullObject:
+    pass
+
+
+class MockShutilWhich:
+    """Mock shutil.which in the tool_meister module."""
+
+    def __init__(self, find: List[str]):
+        """Constructor
+
+        Arguments:
+
+            find:  list of things that should be found
+        """
+        self.find = frozenset(find)
+
+    def which(self, exec: str) -> str:
+        """Mocked version of shutil.which.
+
+        Arguments:
+
+            exec:  A string representing the executable to "look for"
+
+        Returns a "/usr/bin/{exec}" if the exectuable is in the list of things
+        to be found, otherwise None.
+        """
+        return f"/usr/bin/{exec}" if exec in self.find else None
+
+
+def mock_shutil_which(monkeypatch, find: List[str] = []):
```

Comment:
Yeah, my IDE also suggest me to do so :smile:

Reply:
The reason is that the default value is instantiated when the module is _loaded_, and it **persists**; thus, if the function modifies it on one call, then the modification will be present on subsequent calls, and this is liable to be a surprising effect.  So, the way to avoid that is to use a non-mutable object, or to use a flag value (such as `None`) which is replaced by an object which is freshly instantiated on each call to the function.

Reply:
Tracked in #2725.